### PR TITLE
[change] Timeseries data is deleted when device is deleted #419

### LINKED ISSUE #419: [change] Delete tsdb data when device is deleted
Delete tsdb data when device is deleted.

### PR DESCRIPTION
**Pull request by pandafy**
Closes #419

Checks:

- [x] I have manually tested the proposed changes
- [x] I have written new test cases to avoid regressions (if necessary)
- [ ] I have updated the documentation (e.g. README.rst)

### CODE REVIEW

File: openwisp_monitoring/monitoring/base/models.py [hunk 1]
```diff
@@ -115,6 +115,10 @@ def full_clean(self, *args, **kwargs):
         self.key = self._makekey(self.key)
         return super().full_clean(*args, **kwargs)
 
+    @classmethod
+    def post_delete_receiver(cls, instance, *args, **kwargs):
+        timeseries_db.delete_metric_data(key=instance.key, tags=instance.tags)
```
Comment (nemesisdesign) [inline openwisp_monitoring/monitoring/base/models.py:120]: The problem with this is that this may fail, moreover if multiple devices are deleted, the likelyhood of failure is greater.

We need to execute this in a background task with some retry mechanism.
It seems to me that this particular method `delete_metric_data`, does not have a built-in retry mechanism, therefore we could simply let the celery task raise the exception on failures and let the celery task retry whenever the exception is raised, as we do for writing the data.

To sum up, I believe deleting data should work in a similar manner to how the data is written, I think we can simply replicate that kind of structure (executed in a dedicated celery task, which has the retry on failure argument specified).

What do you think?

**Review by nemesisdesign — CHANGES_REQUESTED** [review]

File: openwisp_monitoring/monitoring/base/models.py (hunk 1 above)
Comment (pandafy) [reply openwisp_monitoring/monitoring/base/models.py]: I have added a celery task for deletion of the data. The celery task does not have retry configured because the underlying method already has a retry mechanism. 

https://github.com/openwisp/openwisp-monitoring/blob/e563520c7765d4e638f8e65b8b8e4095fa2fc4cc/openwisp_monitoring/db/backends/influxdb/client.py#L194-L196

File: openwisp_monitoring/monitoring/tasks.py [hunk 2]
```diff
@@ -36,6 +36,11 @@ def timeseries_write(
         post_metric_write.send(**signal_kwargs)
 
 
+@shared_task
+def delete_device_tsdb_data(key, tags):
```
Comment (nemesisdesign) [inline openwisp_monitoring/monitoring/tasks.py:40]: Why are you splitting some code in `openwisp_monitoring.device` and some other lines in `openwisp_monitoring.monitoring`?

It could make sense to have a generic way to remove timeseries data asynchronously, I like the idea, but this task shouldn't have `device` in its name, I'd rename it simply `timeseries_delete`, which is also more consistent with the naming used in the method above.

**Review by nemesisdesign — CHANGES_REQUESTED** [review]
Good progress, see below.

**Review by nemesisdesign — APPROVED** [review]
